lib/datadog/core/diagnostics/health.rb

```diff
@@ -30,6 +37,26 @@ class Metrics < Core::Metrics::Client
           gauge :queue_max_length, Ext::Health::Metrics::METRIC_QUEUE_MAX_LENGTH
           gauge :queue_spans, Ext::Health::Metrics::METRIC_QUEUE_SPANS
           gauge :sampling_service_cache_length, Ext::Health::Metrics::METRIC_SAMPLING_SERVICE_CACHE_LENGTH
+
+          def default_metric_options
+            return super unless @tags
+
+            # Return dupes, so that the constant isn't modified,
+            # and defaults are unfrozen for mutation in Statsd.
+            super.tap do |options|
+              options[:tags] = options[:tags].dup
+              options[:tags] << @tags
+            end
+          end
+
+          private
+
+          # Cache tag strings to avoid recreating them on every flush operation.
+          def compile_tags!
+            return unless @service
+
+            "#{Environment::Ext::TAG_SERVICE}:#{@service}".freeze
+          end
```

Comment:
Hmm is it me or is this code looking deceptively generic, but not actually? E.g. from a high level, there's `@tags = compile_tags!`, which looks generic.

But then `compile_tags!` is not a list but a single tag; and `compile_tags!` doesn't even run if no `@service` is around.

I would suggest picking either making this specific (use a `@service_tag`) or really generic (actually have a list in `compile_tags!`).

(I actually somewhat favor just doing `@service_tag = ...` in the constructor and then referring it directly in `#default_metric_options`).

Reply:
I think the chain of condition introduce unnecessary cognitive burden and this method does not deserved to be extracted private method, since it is just a simple conditional being invoke only once during initialisation.